### PR TITLE
fix(TCOMP-586): Close ClassLoaders that are not reused.

### PR DESCRIPTION
**Pull request by RyanSkraba**
**Please check if the PR fulfills these requirements**

- [X] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format) ?
- [X] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [ ] The code coverage on new code >75%
- [ ] The new code does not introduce new technical issues (sonar / eslint)

**What kind of change does this PR introduce?**

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

ClassLoader caching can be selectively deactivated (see TCOMP-422).  However, when ClassLoaders are not put in the cache, they are not correctly closed after use.

**What is the new behavior?**

Close uncached ClassLoaders when their SandboxedInstance is finished being used.

**Does this PR introduce a breaking change?**

- [ ] Yes
- [X] No

### CODE REVIEW

File: daikon/src/main/java/org/talend/daikon/sandbox/SandboxedInstance.java [hunk 1]
```diff
@@ -46,10 +46,14 @@
 
     private boolean isClosed;
 
-    SandboxedInstance(String classToInstanciate, boolean useCurrentJvmProperties, ClassLoader sandboxClassLoader) {
+    private final boolean reusableClassLoader;
+
+    SandboxedInstance(String classToInstanciate, boolean useCurrentJvmProperties, ClassLoader sandboxClassLoader,
+            boolean reusableClassLoader) {
         this.classToInstanciate = classToInstanciate;
```
Comment (francisu) [inline daikon/src/main/java/org/talend/daikon/sandbox/SandboxedInstance.java]: I like to define a constant value for the boolean this to make the code more reusable when you read it (looking at the test code for example it's not clear what "true" and "false" mean.

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/106/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/106/)

File: daikon/src/main/java/org/talend/daikon/sandbox/SandboxedInstance.java [hunk 2]
```diff
@@ -68,6 +72,18 @@ public void close() {
...
         } // else getInstance was not called so no need to reset context classloader.
+
+        if (!reusableClassLoader) {
+            ClassLoaderIsolatedSystemProperties.getInstance().stopIsolateClassLoader(sandboxClassLoader);
+            if (sandboxClassLoader instanceof AutoCloseable) {
+                try {
+                    ((AutoCloseable) sandboxClassLoader).close();
+                } catch (Exception e) {
+                    new TalendRuntimeException(CommonErrorCodes.UNEXPECTED_EXCEPTION, e);
+                }
```
Comment (francisu) [inline daikon/src/main/java/org/talend/daikon/sandbox/SandboxedInstance.java]: I think TalendRuntimeException.createUnexpectedException(e) is preferred.

**Comment by francisu** [comment]
I don't see a test case for the bug in question; just a modification of the existing test cases. I would like to see that before accepting this.

**Comment by RyanSkraba** [comment]
@francisu : Can you take a look?

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/107/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/107/)

File: daikon/src/main/java/org/talend/daikon/sandbox/SandboxInstanceFactory.java [hunk 3]
```diff
@@ -28,6 +28,12 @@
  */
 public class SandboxInstanceFactory {
 
+    /** A useful constant value for ClassLoaders that can be reused and placed in the cache. */
+    public static final boolean CLASSLOADER_REUSABLE = true;
+
+    /** A useful constant value for ClassLoaders that can not be reused or retrieved from the cache. */
+    public static final boolean CLASSLOADER_NOT_REUSABLE = !CLASSLOADER_REUSABLE;
+
```
Comment (francisu) [inline daikon/src/main/java/org/talend/daikon/sandbox/SandboxInstanceFactory.java]: OK, the 2nd one is not necessary. People know what a ! sign does.

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/108/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/108/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/109/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/109/)

**Pull request merged**